### PR TITLE
Fix H5Dchunk_iter doxygen example, add doxygen example to tests

### PR DESCRIPTION
**Pull request by mkitti**
This updates the doxygen example for `H5Dchunk_iter` following changes made in #2074 .

The primary goal is to change the `chunk_cb` function to the following signature:
```
chunk_cb(const hsize_t *offset, unsigned filter_mask, haddr_t addr, hsize_t size, void *op_data)
```

To ensure the doxygen examples can be compiled, the cmake infrastructure is extended in a limited way to now cover `doxygen/examples/H5D_examples.c`. I anticipate another pull request may extend this further to cover other files in `doxygen/examples`.

### CODE REVIEW

**Comment by mkitti** [comment]
It turns out the doxygen examples have some gcc-isms. Addressing these issues may be too much to deal with in this pull request.

> D:\a\hdf5\hdf5\doxygen\examples\H5D_examples.c(13,12): warning C4477: 'printf' : format string '%ld' requires an argument of type 'long', but variadic argument 1 has type 'hsize_t' [D:\a\hdf5\build\doxygen\examples\H5D_examples.vcxproj]
D:\a\hdf5\hdf5\doxygen\examples\H5D_examples.c(13,12): message : consider using '%lld' in the format string [D:\a\hdf5\build\doxygen\examples\H5D_examples.vcxproj]
D:\a\hdf5\hdf5\doxygen\examples\H5D_examples.c(13,12): message : consider using '%Id' in the format string [D:\a\hdf5\build\doxygen\examples\H5D_examples.vcxproj]
D:\a\hdf5\hdf5\doxygen\examples\H5D_examples.c(13,12): message : consider using '%I64d' in the format string [D:\a\hdf5\build\doxygen\examples\H5D_examples.vcxproj]
D:\a\hdf5\hdf5\doxygen\examples\H5D_examples.c(38,23): error C2065: '__label__': undeclared identifier [D:\a\hdf5\build\doxygen\examples\H5D_examples.vcxproj]
D:\a\hdf5\hdf5\doxygen\examples\H5D_examples.c(38,23): error C2146: syntax error: missing ';' before identifier 'fail_dtype' [D:\a\hdf5\build\doxygen\examples\H5D_examples.vcxproj]
D:\a\hdf5\hdf5\doxygen\examples\H5D_examples.c(38,33): error C2065: 'fail_dtype': undeclared identifier [D:\a\hdf5\build\doxygen\examples\H5D_examples.vcxproj]
D:\a\hdf5\hdf5\doxygen\examples\H5D_examples.c(38,46): error C2065: 'fail_dspace': undeclared identifier [D:\a\hdf5\build\doxygen\examples\H5D_examples.vcxproj]
D:\a\hdf5\hdf5\doxygen\examples\H5D_examples.c(38,58): error C2065: 'fail_shape': undeclared identifier [D:\a\hdf5\build\doxygen\examples\H5D_examples.vcxproj]
D:\a\hdf5\hdf5\doxygen\examples\H5D_examples.c(64,20): warning C4477: 'printf' : format string '%lu' requires an argument of type 'unsigned long', but variadic argument 3 has type 'size_t' [D:\a\hdf5\build\doxygen\examples\H5D_examples.vcxproj]
D:\a\hdf5\hdf5\doxygen\examples\H5D_examples.c(64,20): message : consider using '%zu' in the format string [D:\a\hdf5\build\doxygen\examples\H5D_examples.vcxproj]
D:\a\hdf5\hdf5\doxygen\examples\H5D_examples.c(95,19): error C2065: '__label__': undeclared identifier [D:\a\hdf5\build\doxygen\examples\H5D_examples.vcxproj]
D:\a\hdf5\hdf5\doxygen\examples\H5D_examples.c(95,19): error C2146: syntax error: missing ';' before identifier 'fail_lcpl' [D:\a\hdf5\build\doxygen\examples\H5D_examples.vcxproj]
D:\a\hdf5\hdf5\doxygen\examples\H5D_examples.c(95,28): error C2065: 'fail_lcpl': undeclared identifier [D:\a\hdf5\build\doxygen\examples\H5D_examples.vcxproj]
D:\a\hdf5\hdf5\doxygen\examples\H5D_examples.c(95,39): error C2065: 'fail_dset': undeclared identifier [D:\a\hdf5\build\doxygen\examples\H5D_examples.vcxproj]
D:\a\hdf5\hdf5\doxygen\examples\H5D_examples.c(95,50): error C2065: 'fail_file': undeclared identifier [D:\a\hdf5\build\doxygen\examples\H5D_examples.vcxproj]
D:\a\hdf5\hdf5\doxygen\examples\H5D_examples.c(146,19): error C2065: '__label__': undeclared identifier [D:\a\hdf5\build\doxygen\examples\H5D_examples.vcxproj]
D:\a\hdf5\hdf5\doxygen\examples\H5D_examples.c(146,19): error C2146: syntax error: missing ';' before identifier 'fail_dset' [D:\a\hdf5\build\doxygen\examples\H5D_examples.vcxproj]
D:\a\hdf5\hdf5\doxygen\examples\H5D_examples.c(146,28): error C2065: 'fail_dset': undeclared identifier [D:\a\hdf5\build\doxygen\examples\H5D_examples.vcxproj]
D:\a\hdf5\hdf5\doxygen\examples\H5D_examples.c(146,39): error C2065: 'fail_file': undeclared identifier [D:\a\hdf5\build\doxygen\examples\H5D_examples.vcxproj]
D:\a\hdf5\hdf5\doxygen\examples\H5D_examples.c(178,19): error C2065: '__label__': undeclared identifier [D:\a\hdf5\build\doxygen\examples\H5D_examples.vcxproj]
D:\a\hdf5\hdf5\doxygen\examples\H5D_examples.c(178,19): error C2146: syntax error: missing ';' before identifier 'fail_update' [D:\a\hdf5\build\doxygen\examples\H5D_examples.vcxproj]
D:\a\hdf5\hdf5\doxygen\examples\H5D_examples.c(178,30): error C2065: 'fail_update': undeclared identifier [D:\a\hdf5\build\doxygen\examples\H5D_examples.vcxproj]
D:\a\hdf5\hdf5\doxygen\examples\H5D_examples.c(178,43): error C2065: 'fail_fspace': undeclared identifier [D:\a\hdf5\build\doxygen\examples\H5D_examples.vcxproj]
D:\a\hdf5\hdf5\doxygen\examples\H5D_examples.c(178,54): error C2065: 'fail_dset': undeclared identifier [D:\a\hdf5\build\doxygen\examples\H5D_examples.vcxproj]
D:\a\hdf5\hdf5\doxygen\examples\H5D_examples.c(178,65): error C2065: 'fail_file': undeclared identifier [D:\a\hdf5\build\doxygen\examples\H5D_examples.vcxproj]
D:\a\hdf5\hdf5\doxygen\examples\H5D_examples.c(222,19): error C2065: '__label__': undeclared identifier [D:\a\hdf5\build\doxygen\examples\H5D_examples.vcxproj]
D:\a\hdf5\hdf5\doxygen\examples\H5D_examples.c(222,19): error C2146: syntax error: missing ';' before identifier 'fail_delete' [D:\a\hdf5\build\doxygen\examples\H5D_examples.vcxproj]
D:\a\hdf5\hdf5\doxygen\examples\H5D_examples.c(222,30): error C2065: 'fail_delete': undeclared identifier [D:\a\hdf5\build\doxygen\examples\H5D_examples.vcxproj]
D:\a\hdf5\hdf5\doxygen\examples\H5D_examples.c(222,41): error C2065: 'fail_file': undeclared identifier [D:\a\hdf5\build\doxygen\examples\H5D_examples.vcxproj]
D:\a\hdf5\hdf5\doxygen\examples\H5D_examples.c(170,1): error C2045: 'fail_dset': label redefined [D:\a\hdf5\build\doxygen\examples\H5D_examples.vcxproj]
D:\a\hdf5\hdf5\doxygen\examples\H5D_examples.c(172,1): error C2045: 'fail_file': label redefined [D:\a\hdf5\build\doxygen\examples\H5D_examples.vcxproj]
D:\a\hdf5\hdf5\doxygen\examples\H5D_examples.c(212,1): error C2045: 'fail_fspace': label redefined [D:\a\hdf5\build\doxygen\examples\H5D_examples.vcxproj]
D:\a\hdf5\hdf5\doxygen\examples\H5D_examples.c(214,1): error C2045: 'fail_dset': label redefined [D:\a\hdf5\build\doxygen\examples\H5D_examples.vcxproj]
D:\a\hdf5\hdf5\doxygen\examples\H5D_examples.c(216,1): error C2045: 'fail_file': label redefined [D:\a\hdf5\build\doxygen\examples\H5D_examples.vcxproj]
D:\a\hdf5\hdf5\doxygen\examples\H5D_examples.c(247,1): error C2045: 'fail_file': label redefined [D:\a\hdf5\build\doxygen\examples\H5D_examples.vcxproj]

**Review by byrnHDF — APPROVED** [review]

**Comment by mkitti** [comment]
Would it be easier to have a pull request that just did the changes to `H5D_examples.c`?

File: doxygen/examples/CMakeTests.cmake [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+# If you do not have access to either file, you may request a copy from
+# help@hdfgroup.org.
+#
+
+##############################################################################
+##############################################################################
+###           T E S T I N G                                                ###
+##############################################################################
+##############################################################################
+file (MAKE_DIRECTORY ${PROJECT_BINARY_DIR}/red ${PROJECT_BINARY_DIR}/blue ${PROJECT_BINARY_DIR}/u2w)
```
Comment (byrnHDF) [inline doxygen/examples/CMakeTests.cmake]: Does the example expect these dirs?

File: doxygen/examples/CMakeTests.cmake [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+          -D "TEST_ARGS:STRING="
+          -D "TEST_EXPECT=0"
+          -D "TEST_SKIP_COMPARE=TRUE"
+          -D "TEST_OUTPUT=${example}.txt"
+          -D "TEST_FOLDER=${PROJECT_BINARY_DIR}"
+          -P "${HDF_RESOURCES_DIR}/runTest.cmake"
+      )
+    endif ()
+    set_tests_properties (DOXYGEN-EXAMPLES-${example} PROPERTIES FIXTURES_REQUIRED clear_DOXYGEN_EXAMPLES)
+    if (last_test)
```
Comment (byrnHDF) [inline doxygen/examples/CMakeTests.cmake]: we shouldn't need last_test variable unless there is a sequence dependency.

File: doxygen/examples/CMakeTests.cmake (hunk 1 above)
Comment (mkitti) [reply doxygen/examples/CMakeTests.cmake]: Honestly, I believe I copied this file over from `./examples/CMakeTests.cmake`, so this may unnecessary here.

**Comment by mkitti** [comment]
Now that I've extracted the changes to `H5D_examples.c` out, perhaps this branch should focus on testing all of the doxygen documentation.

File: doxygen/examples/CMakeTests.cmake (hunk 2 above)
Comment (mkitti) [reply doxygen/examples/CMakeTests.cmake]: Fixed in [12fbb84](https://github.com/HDFGroup/hdf5/pull/2380/commits/12fbb846aae95b34806dadf56507c87b647bf373)

File: doxygen/examples/CMakeTests.cmake (hunk 1 above)
Comment (mkitti) [reply doxygen/examples/CMakeTests.cmake]: Fixed in [12fbb84](https://github.com/HDFGroup/hdf5/pull/2380/commits/12fbb846aae95b34806dadf56507c87b647bf373)

**Comment by mkitti** [comment]
In https://github.com/HDFGroup/hdf5/pull/2329#issuecomment-1440891847, jhendersonHDF recommended the use of the PRIuHSIZE macro.

**Comment by mkitti** [comment]
I believe the H5DO_examples.c are completely out of date <strike>and it is not clear to me that this is even used in the Doxygen documentation</strike>. Perhaps it should just be removed?

`H5DOwrite_chunk` and `H5DOread_chunk` were renamed a while ago.

**Review by derobins — APPROVED** [review]

**Comment by bmribler** [comment]
> I believe the H5DO_examples.c are completely out of date and it is not clear to me that this is even used in the Doxygen documentation. Perhaps it should just be removed?
> 
> `H5DOwrite_chunk` and `H5DOread_chunk` were renamed a while ago.

Do you plan to create an issue for updating H5DO_examples.c, @mkitti?

**Comment by mkitti** [comment]
What's the direction? Do we retool the examples to use the current names of these functions?

**Comment by byrnHDF** [comment]
> What's the direction? Do we retool the examples to use the current names of these functions?

Two things - we need to use the hdf5-examples repo by transferring it into hdf5 repo and eliminate other examples. Then we would have one source that is proper usage and update/add neew examples in one place.

**Comment by mkitti** [comment]
Can we import general code into Doxygen as examples?

**Comment by byrnHDF** [comment]
> Can we import general code into Doxygen as examples?

Not sure what you mean, but if we use the HDF5-examples, doxygen would need to add those files - Probably we should even add a separate page/section of just the examples.  Then using the snip markdown we can import sections anywhere.  (Note that I mean to replace all the examples with the code from the repo literally. Then the repo could be retired.)

**Comment by mkitti** [comment]
I'm getting a bit lost of where to go from here. I'm going to close for the moment since it seems starting from scratch might be the easiest approach. Feel free to take the initiative if I do not get back to this first.

The main objective is to make sure that all examples in Doxygen can actually compile correctly without warnings.

**Pull request closed without merging**